service/storage_service.cc

```diff
@@ -6079,12 +6087,32 @@ future<> storage_service::stream_tablet(locator::global_tablet_id tablet) {
         auto range = tmap.get_token_range(tablet.tablet);
         std::optional<locator::tablet_replica> leaving_replica = locator::get_leaving_replica(tinfo, *trinfo);
         locator::tablet_migration_streaming_info streaming_info = get_migration_streaming_info(tm->get_topology(), tinfo, *trinfo);
+        locator::tablet_replica_set read_from{streaming_info.read_from.begin(), streaming_info.read_from.end()};
+        if (trinfo->transition == locator::tablet_transition_kind::rebuild_v2) {
+            read_from = read_from | std::views::filter([&pending_replica = *trinfo->pending_replica] (const auto& tr) {
+                return tr != pending_replica;
+            }) | std::ranges::to<locator::tablet_replica_set>();
+
+            auto nearest_hosts = read_from | std::views::transform([] (const auto& tr) {
+                return tr.host;
+            }) | std::ranges::to<host_id_vector_replica_set>();
+            tm->get_topology().sort_by_proximity(trinfo->pending_replica->host, nearest_hosts);
+
+            if (!nearest_hosts.empty()) {
+                auto it = std::find_if(read_from.begin(), read_from.end(), [nearest_host = nearest_hosts[0]] (const auto& tr) { return tr.host == nearest_host; });
```

Comment:
Is this enough

assert streaming_info.read_from.contains(nearest_hosts[0] )

?

Reply:
read_from is a tablet_replica set, not host_id set. This code is looking up tablet_replica on the nearest host.